src/analysis/normed/ring/seminorm_constructions.lean

```diff
@@ -0,0 +1,339 @@
+import analysis.normed.ring.seminorm
+import topology.metric_space.basic
+import topology.algebra.order.monotone_convergence
+
+/-!
+# Smoothing procedures for seminorms
+
+This file contains several ring seminorm constructions that will be used in the proof that a rank 1
+nonarchimedean valuation on a complete field can be uniquely extended to any algebraic extension.
+
+## Main declarations
+
+* `seminorm_from_const`: Given a power-multiplicative ring seminorm `f` on a ring `R` and an
+  element `c ∈ R` with `f c ≠ 0`, the real-valued function sending `x ∈ R` to the limit of
+  `(f (x * c^n))/((f c)^n)` is a power-multiplicative ring seminorm on `R` satisfying some
+  multiplicativity conditions [BGR84, Proposition 1.3.2/2].
+
+
+## References
+
+* [S. Bosch, U. Güntzer, R. Remmert, *Non-Archimedean Analysis*][bosch-guntzer-remmert]
+
+## Tags
+ring_seminorm, ring_norm
+-/
+
+
+noncomputable theory
+
+open_locale topological_space
+
+section defs
+
+/-- A function `f : α → β` is power-multiplicative if for all `a ∈ α` and all positive `n ∈ ℕ`,
+  `f (a ^ n) = (f a) ^ n`. -/
+def is_pow_mul {α : Type*} [has_pow α ℕ] {β : Type*} [has_pow β ℕ]  (f : α → β) :=
+∀ (a : α) {n : ℕ} (hn : 1 ≤ n), f (a ^ n) = (f a) ^ n
+
+/-- A function `f : α → β` is nonarchimedean if it satisfies the inequality
+  `f (a + b) ≤ max (f a) (f b)` for all `a, b ∈ α`. -/
+def is_nonarchimedean {α : Type*} [has_add α] {β : Type*} [linear_order β] (f : α → β) : Prop :=
+∀ r s, f (r + s) ≤ max (f r) (f s)
+
+end defs
+
+section seminorm_from_const
+/- In this section we prove Proposition 1.3.2/2 in
+[S. Bosch, U. Güntzer, R. Remmert, *Non-Archimedean Analysis*][bosch-guntzer-remmert]. -/
+
+section ring
+
+variables {R : Type*} [comm_ring R] (c : R) (f : ring_seminorm R) (hf1 : f 1 ≤ 1)
+  (hc : 0 ≠ f c) (hpm : is_pow_mul f)
+
+/-- For a ring seminorm `f` on `R` and `c ∈ R`, the sequence given by `(f (x * c^n))/((f c)^n)`. -/
+def seminorm_from_const_seq (x : R) : ℕ → ℝ :=
+λ n, (f (x * c^n))/((f c)^n)
+
+lemma seminorm_from_const_nonneg (x : R) (n : ℕ) : 0 ≤ seminorm_from_const_seq c f x n :=
+div_nonneg (map_nonneg f (x * c ^ n)) (pow_nonneg (map_nonneg f c) n)
+
+lemma seminorm_from_const_is_bdd_below (x : R) :
+  bdd_below (set.range (seminorm_from_const_seq c f x)) :=
+begin
+  use 0,
+  rw mem_lower_bounds,
+  rintros r ⟨n, hn⟩,
+  rw ← hn,
+  exact seminorm_from_const_nonneg c f x n,
+end
+
+variable {f}
+
+lemma seminorm_from_const_seq_zero (hf : f 0 = 0) :
+  seminorm_from_const_seq c f 0 = 0 :=
+begin
+  simp only [seminorm_from_const_seq],
+  ext n,
+  rw [zero_mul, hf, zero_div],
+  refl,
+end
+
+variable {c}
+
+include hc hpm
+
+lemma seminorm_from_const_seq_one (n : ℕ) (hn : 1 ≤ n) :
+  seminorm_from_const_seq c f 1 n = 1 :=
+begin
+  simp only [seminorm_from_const_seq],
+  rw [one_mul, hpm _ hn, div_self (pow_ne_zero n (ne.symm hc))],
+end
+
+include hf1
+
+lemma seminorm_from_const_seq_antitone (x : R) :
+  antitone (seminorm_from_const_seq c f x) :=
+begin
```

Comment:
I think there's an `antitone_nat_of_succ_le` lemma or similar that you can use to golf this.

Reply:
Thanks!